Dependency parser doesn't respect ecmaFeatures

This style of code is very common in projects using React and higher order components. This rule doesn't seem to detect the default export in this case.

    //connect a redux store
    export default connect()(App);

Ah sorry, it actually appears that JSX is the cause.
My classes with JSX were wrapped in connect calls and didn't work but my classes with standard exports worked fine.

App.js
    
    export default class App {
      render() {
        return <h1>Hello World!</h1>; //comment this line out and it works
      }
    }
    
index.js

    import App from './App';

.eslintrc

    {
      "root": true,

      "env": {
        "es6": true
      },
    
      "ecmaFeatures": {
        "classes": true,
        "modules": true,
        "jsx": true
      },
    
      "plugins": [
        "import"
      ],
    
      "rules": {
        "import/default": 2
      }
    
    }

eslint ./index.js

    /Users/james/demo/index.js
      1:8  error  No default export found in module  import/default

--- a/src/core/getExports.js
+++ b/src/core/getExports.js
@@ -6,12 +6,14 @@ import isIgnored from './ignore'
 var exportCache = new Map()
 
 export default class ExportMap {
-  constructor(settings) {
-    this.settings = settings
+  constructor(context) {
+    this.context = context
     this.named = new Set()
 
     this.errors = []
   }
+
+  get settings() { return this.context && this.context.settings }
 
   get hasDefault() { return this.named.has('default') }
   get hasNamed() { return this.named.size > (this.hasDefault ? 1 : 0) }
@@ -23,14 +25,14 @@ export default class ExportMap {
     var path = resolve(source, context)
     if (path == null || isIgnored(path, context)) return null
 
-    return ExportMap.for(path, context.settings)
+    return ExportMap.for(path, context)
   }
 
-  static for(path, settings) {
+  static for(path, context) {
     var exportMap = exportCache.get(path)
     if (exportMap != null) return exportMap
 
-    exportMap = ExportMap.parse(path, settings)
+    exportMap = ExportMap.parse(path, context)
 
     exportCache.set(path, exportMap)
 
@@ -40,11 +42,11 @@ export default class ExportMap {
     return exportMap
   }
 
-  static parse(path, settings) {
-    var m = new ExportMap(settings)
+  static parse(path, context) {
+    var m = new ExportMap(context)
 
     try {
-      var ast = parse(path, settings)
+      var ast = parse(path, context)
     } catch (err) {
       m.errors.push(err)
       return m // can't continue
@@ -63,7 +65,7 @@ export default class ExportMap {
     var remotePath = resolve.relative(node.source.value, base, this.settings)
     if (remotePath == null) return null
 
-    return ExportMap.for(remotePath, this.settings)
+    return ExportMap.for(remotePath, this.context)
   }
 
   captureDefault(n) {
@@ -125,3 +127,4 @@ export default class ExportMap {
     }.bind(this))
   }
 }
+//

--- a/src/core/parse.js
+++ b/src/core/parse.js
@@ -5,15 +5,27 @@ const defaultParseOptions = { ecmaVersion: 6  // for espree, esprima. not needed
                             , sourceType: "module"
                             }
 
-export default function parse(path, settings) {
-  const parser = require(settings['import/parser'] || "babylon")
+export default function parse(path, context) {
+  const { settings } = context
+      , parser = settings['import/parser'] || "babylon"
+
+  const { parse } = require(parser)
       , options = Object.assign( {}
                                , defaultParseOptions
                                , settings['import/parse-options'])
 
-  const ast = parser.parse( fs.readFileSync(path, {encoding: 'utf8'})
-                          , options
-                          )
+  // detect and handle "jsx" ecmaFeature
+  if (context.ecmaFeatures && parser === "babylon") {
+    const { jsx } = context.ecmaFeatures
+    if (jsx && (!options.plugins || options.plugins.indexOf('jsx') < 0)) {
+      if (!options.plugins) options.plugins = ['jsx']
+      else options.plugins.push('jsx')
+    }
+  }
+
+  const ast = parse( fs.readFileSync(path, {encoding: 'utf8'})
+                   , options
+                   )
 
   // bablyon returns top-level "File" node.
   return ast.type === 'File' ? ast.program : ast

--- a/src/core/resolve.js
+++ b/src/core/resolve.js
@@ -29,7 +29,7 @@ function fileExists(filepath) {
 function opts(basedir, settings) {
   // pulls all items from 'import/resolve'
   return Object.assign( { }
-                      , settings['import/resolve']
+                      , settings && settings['import/resolve']
                       , { basedir: basedir }
                       )
 }

--- a/tests/files/jsx/App.js
+++ b/tests/files/jsx/App.js
@@ -1,0 +1,9 @@
+import React, { Component } from 'react'
+
+// class App extends Component {
+  // render() {
+    // return <div>hello, JSX</div>
+  // }
+// }
+
+export default connect()(App);

--- a/tests/files/redux.js
+++ b/tests/files/redux.js
@@ -1,0 +1,2 @@
+// redux idiom
+export default connect()(App);

--- a/tests/src/core/getExports.js
+++ b/tests/src/core/getExports.js
@@ -51,7 +51,7 @@ describe('getExports', function () {
 
   it('finds exports for an ES7 module with babel-eslint', function () {
     var imports = ExportMap.parse( getFilename('jsx/FooES7.js')
-                                 , { 'import/parser': 'babel-eslint' })
+                                 , { settings: { 'import/parser': 'babel-eslint' } })
 
     expect(imports).to.exist
     expect(imports).to.have.property('hasDefault', true)
@@ -59,7 +59,7 @@ describe('getExports', function () {
   })
 
   it('finds exports for an ES7 module with proper parse options', function () {
-    var imports = ExportMap.parse(getFilename('jsx/FooES7.js'), {
+    var imports = ExportMap.parse(getFilename('jsx/FooES7.js'), { settings: {
       'import/parse-options': {
         plugins: [
           'decorators',
@@ -68,7 +68,7 @@ describe('getExports', function () {
           'objectRestSpread'
         ]
       }
-    })
+    }})
 
     expect(imports).to.exist
     expect(imports).to.have.property('hasDefault', true)

--- a/tests/src/rules/default.js
+++ b/tests/src/rules/default.js
@@ -41,12 +41,19 @@ ruleTester.run('default', rule, {
   , test({ code: 'import twofer from "./trampoline"'
          , settings: { 'import/parse-options': { plugins: ['exportExtensions']}}
          })
+    // jsx
+  , test({ code: 'import MyCoolComponent from "./jsx/MyCoolComponent.jsx"'
+         , ecmaFeatures: { jsx: true, modules: true }
+         })
 
     // #54: import of named export default
   , test({ code: 'import foo from "./named-default-export"' })
+
     // #94: redux export of execution result
   , test({ code: 'import connectedApp from "./redux"' })
-
+  , test({ code: 'import App from "./jsx/App"'
+         , ecmaFeatures: { jsx: true, modules: true }
+         })
   ],
 
   invalid: [

--- a/tests/src/rules/default.js
+++ b/tests/src/rules/default.js
@@ -44,6 +44,8 @@ ruleTester.run('default', rule, {
 
     // #54: import of named export default
   , test({ code: 'import foo from "./named-default-export"' })
+    // #94: redux export of execution result
+  , test({ code: 'import connectedApp from "./redux"' })
 
   ],
 

